addons/account/models/reconciliation_widget.py

```diff
@@ -75,9 +76,20 @@ def get_move_lines_for_bank_statement_line(self, st_line_id, partner_id=None, ex
         if partner_id is None:
             partner_id = st_line.partner_id.id
 
-        domain = self._domain_move_lines_for_reconciliation(st_line, aml_accounts, partner_id, excluded_ids=excluded_ids, search_str=search_str)
+        domain = self._domain_move_lines_for_reconciliation(st_line, aml_accounts, partner_id, excluded_ids=excluded_ids, search_str=search_str, mode=mode)
         recs_count = self.env['account.move.line'].search_count(domain)
-        aml_recs = self.env['account.move.line'].search(domain, offset=offset, limit=limit, order="date_maturity desc, id desc")
+
+        from_clause, where_clause, where_clause_params = self.env['account.move.line']._where_calc(domain).get_sql()
+        where_str = where_clause and (" WHERE %s" % where_clause) or ''
+        order_by = ''' order by ("account_move_line".debit - "account_move_line".credit) = {amount} desc,
+                      "account_move_line".date_maturity asc,
+                      "account_move_line".id asc
+        '''.format(amount=st_line.amount)
+        query_str = 'SELECT "%s".id FROM ' % self.env['account.move.line']._table + from_clause + where_str + order_by + (limit and ' limit %d' % limit or '') + (offset and ' offset %d' % offset or '')
+        self._cr.execute(query_str, where_clause_params)
```

Comment:
query_str could have been built more elegantly

Reply:
I improved it a tiny bit but I don't know how to do it better

Reply:
what about
```python
        from_clause, where_clause, params = self.env['account.move.line']._where_calc(domain).get_sql()
        query_str = '''
            SELECT "account_move_line".id 
            FROM ''' + from_clause 
        query_str += where_clause and (" WHERE %s" % where_clause) or ''
        query_str += '''
             ORDER BY ("account_move_line".debit - "account_move_line".credit) = {amount} DESC,
               "account_move_line".date_maturity ASC,
               "account_move_line".id ASC
        '''.format(amount=st_line.amount)
        if limit:
            query_str += ' LIMIT %s'
            params += [limit]
        if offset:
            query_str += ' OFFSET %s'
            params += [offset]
        self._cr.execute(query_str, params)
```

note that this also avoid SQL injections on limit and offset params